### PR TITLE
Refactor replication event and resend handler

### PR DESCRIPTION
**Pull request by xwduan**
## What changed?

Refactor replication event handler and resend handler for replication stack
## Why?

To support migration back and forth case
## How did you test it?

Unit test and integration test
## Potential risks

No risk. The code are gated by a dynamic config
## Documentation

N/a
## Is hotfix candidate?

no

### CODE REVIEW

File: service/history/replication/eventhandler/event_importer.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+			if err != nil {
+				return err
+			}
+			blobs = []*common.DataBlob{}
+			blobSize = 0
+			token = response.Token
+		}
+	}
+	if len(blobs) != 0 {
+		response, err := invokeImportWorkflowExecutionCall(ctx, engine, workflowKey, blobs, versionHistory, token, e.logger)
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/eventhandler/event_importer.go:162]: Updated the code to check the version before adding blob. So we don't need to check here.

File: service/history/replication/eventhandler/event_importer.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+		}
+
+		if versionHistory != nil && !versionhistory.IsVersionHistoryItemsInSameBranch(versionHistory.Items, batch.VersionHistory.Items) {
+			return serviceerror.NewInternal("History Branch changed during importing")
+		}
+		events, err := e.serializer.DeserializeEvents(batch.RawEventBatch)
+		if err != nil {
+			return err
+		}
+		if len(events) == 0 {
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/eventhandler/event_importer.go:141]: Current [implementation](https://github.com/temporalio/temporal/blob/main/common/xdc/ndc_history_resender.go) does not handle this case at all. so I am hesitating to return error in this case. I am not sure if the api could return empty batch or not. I will change it to return error. If empty batch returned, we should fix the api.

File: service/history/replication/eventhandler/event_importer.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+			continue
+		}
+		if eventsVersion != common2.EmptyVersion && eventsVersion != events[0].GetVersion() {
+			if err := importFn(); err != nil {
+				return err
+			}
+		}
+		versionHistory = batch.VersionHistory
+		eventsVersion = events[0].GetVersion()
+		blobSize++
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/eventhandler/event_importer.go:151]: good catch. it should be `blobSize += len(blob.Data)`

File: service/history/replication/eventhandler/event_importer.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+		blobs = append(blobs, batch.RawEventBatch)
+
+		if blobSize >= historyImportBlobSize || len(blobs) >= historyImportPageSize {
+			if err := importFn(); err != nil {
+				return err
+			}
+		}
+	}
+	if len(blobs) != 0 {
+		response, err := invokeImportWorkflowExecutionCall(ctx, engine, workflowKey, blobs, versionHistory, token, e.logger)
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/eventhandler/event_importer.go:161]: Sure.

File: service/history/replication/eventhandler/remote_history_paginated_fetcher.go [hunk 5]
```diff
@@ -109,21 +120,34 @@ func (n *HistoryPaginatedFetcherImpl) GetSingleWorkflowHistoryPaginatedIterator(
...
-	resendCtx := context.Background()
-	var cancel context.CancelFunc
-	if n.rereplicationTimeout != nil {
-		resendContextTimeout := n.rereplicationTimeout(namespaceID.String())
-		if resendContextTimeout > 0 {
-			resendCtx, cancel = context.WithTimeout(resendCtx, resendContextTimeout)
-			defer cancel()
-		}
-	}
-	resendCtx = rpc.CopyContextValues(resendCtx, ctx)
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/eventhandler/remote_history_paginated_fetcher.go:123]: It is moved to line 210.

**Review by hai719 — APPROVED** [review]

**Pull request merged**